### ISSUE TITLE
Inconsistent symbol duplication checking

### ISSUE THREAD

**Issue by vadimcn** [rust-lang/rust#21546]
This compiles:
```rust
mod Foo {}
struct Foo;

fn main() {}
```
but this doesn't:
```rust
mod Foo {}
struct Bar;
type Foo = Bar;

fn main() {}
```

```
<anon>:3:1: 3:16 error: duplicate definition of type or module `Foo`
<anon>:3 type Foo = Bar;
         ^~~~~~~~~~~~~~~
<anon>:1:1: 1:11 note: first definition of type or module `Foo` here
<anon>:1 mod Foo {}
         ^~~~~~~~~~
```
  
I'd have expected both to either succeed or fail.

**Comment by nham** [comment]
This shouldn't be closed yet. #26421 didn't fix the issue, it only added warnings for it so it can be fixed later.

**Comment by nrc** [comment]
Fixed by #29530
